test/example-72-function-attrs-with-macros.rst

```diff
@@ -0,0 +1,24 @@
+
+.. c:function:: int hidden_frob(struct list * list, enum mode mode)
+
+   List frobnicator with visibility hidden attribute.
+
+   :param list: The list to frob.
+   :param mode: The frobnication mode.
+   :return: 0 on success, non-zero error code on error.
+   :since: v0.1
+
+**Visibility** hidden
```

Comment:
I haven't checked the compiled output, but assuming this looks fine in the compiled documentation I have only one other comment: I'd like to see `**EXPORT_HIDEN**` instead and then you can document the macro itself if you want. That was the whole point I tried to make with the `__init` macro. Note that the macro also carries a lot more meaning in the context of a such a project in my opinion. Maybe I'm alone on this one though...

A more interesting question is whether that would mean that we should also get `__attribute__((pure))` verbatim in the other example. I know it's uglier, but I'm trying to find common ground rather than hard coding things that are specific to how one project uses them. And the common ground is to do as little as possible.

Note I haven't looked at Clang's AST for this and also, what would Clang put there if the define was elsewhere and not reachable when generating the docs? I believe Clang won't include other header files in the way we use it, but I haven't touched this in a couple of months, I may be forgetting stuff.

Reply:
> I have only one other comment: I'd like to see **EXPORT_HIDEN** instead and then you can document the macro itself if you want. 

Placing the macro name directly means to extract it from the tokens comments. As the C-preprocessor is ran it will replace the macro with the attribute itself so when the parser goes over the AST it won't be present. I for one would've like to see it present as well, though, in this context, putting in front of the symbol wouldn't do any justice (as it isn't a return type). Doing this would probably violate one of the design principles of hawkmoth?

> I haven't checked the compiled output, but assuming this looks fine in the compiled documentation

It looks pretty OK to me: https://people.collabora.com/~mvlad/libweston-hawkmoth-group/usage/input.html#this-displays-only-functions-in-group-seat

> A more interesting question is whether that would mean that we should also get __attribute__((pure)) verbatim in the other example. I know it's uglier, but I'm trying to find common ground rather than hard coding things that are specific to how one project uses them. And the common ground is to do as little as possible.

In my "experiments" not all attributes are "equal". Some like (`const`, `pure`, `visibility`) are tagged as a node in the AST. Others like `noreturn` are in the function body but not as return types: 

```
`-FunctionDecl 0x350d4f0 <line:3:18, line:36:51> col:6 no_ret_frob 'void (struct list *, enum mode) __attribute__((noreturn))'
  |-ParmVarDecl 0x350d1e0 <col:18, col:31> col:31 list 'struct list *'
  |-ParmVarDecl 0x350d358 <col:37, col:47> col:47 mode 'enum mode':'enum mode'
  `-FullComment 0x350da20 <line:28:3, line:33:15>
    |-ParagraphComment 0x350d930 <line:28:3, col:44>
    | `-TextComment 0x350d900 <col:3, col:44> Text=" List frobnicator with noreturn attribute."
    `-ParagraphComment 0x350d9f0 <line:30:3, line:33:15>
      |-TextComment 0x350d950 <line:30:3, col:33> Text=" :param list: The list to frob."
      |-TextComment 0x350d970 <line:31:3, col:38> Text=" :param mode: The frobnication mode."
      |-TextComment 0x350d990 <line:32:3, col:55> Text=" :return: 0 on success, non-zero error code on error."
      `-TextComment 0x350d9b0 <line:33:3, col:15> Text=" :since: v0.1"
```

So to get back to your questions: you won't see `pure` attribute in the body just in the node as Pure attribute:

```
|-FunctionDecl 0x256e328 <test/example-71-function-attrs.c:9:1, line:10:48> col:5 pure_frob 'int (struct list *, enum mode)'
| |-ParmVarDecl 0x256e090 <col:15, col:28> col:28 list 'struct list *'
| |-ParmVarDecl 0x256e208 <col:34, col:44> col:44 mode 'enum mode':'enum mode'
| |-PureAttr 0x256e3d8 <line:9:16>
| `-FullComment 0x2596d50 <line:2:3, line:7:15>
|   |-ParagraphComment 0x2596c60 <line:2:3, col:40>
|   | `-TextComment 0x2596c30 <col:3, col:40> Text=" List frobnicator with pure attribute."
|   `-ParagraphComment 0x2596d20 <line:4:3, line:7:15>
|     |-TextComment 0x2596c80 <line:4:3, col:33> Text=" :param list: The list to frob."
|     |-TextComment 0x2596ca0 <line:5:3, col:38> Text=" :param mode: The frobnication mode."
|     |-TextComment 0x2596cc0 <line:6:3, col:55> Text=" :return: 0 on success, non-zero error code on error."
|     `-TextComment 0x2596ce0 <line:7:3, col:15> Text=" :since: v0.1"
```

Reply:
ping @BrunoMSantos would be nice to get some feedback here as well.

Reply:
Sorry, day job gets in the way, I'm planning some more serious follow ups over the weekend :)

Yeah, this one's a bugger. As a user I actually don't care because I would never add this to the documentation myself (hence my request for this being optional), but as a developer interested in making it work for everyone who may want it, I still think my points are valid.

All I can say is that your answer is very detailed and raises interesting points, but it only tells me this can't be done right --- i.e. simple/robust and universal. I'd be willing to sacrifice simplicity if we achieved universality, and I mean that in 2 ways: in the sense that it handles any attribute or macro the same way and in that it makes no assumptions about what they might mean, right down to the expansion of macros (`__init` reads totally different from say 'place this code in cold section, optimize for space, etc.'). If you can find a way of doing that, I may be willing to support the idea myself.

I also think that some of your goals in these PRs are interdependent and currently biased towards your existing Doxygen documentation. Have you considered if any of the suggested solutions would work for you? I.e. splitting API into separate headers (or separate it with new directives or whatever comes out of the grouping PR) and / or documenting in plain English if a certain function is part of the API or not.

If not, I think it's best if I back down and let the dictator (@jnikula) decide. I really have no further input at the design level :/

Reply:
> I also think that some of your goals in these PRs are interdependent and currently biased towards your existing Doxygen documentation.